tests/vu/lower/efu.cpp

```diff
@@ -0,0 +1,155 @@
+static const u32 __attribute__((aligned(16))) C_NEGONE[4] = {0xBF800000, 0xBF800000, 0xBF800000, 0xBF800000};
+static const u32 __attribute__((aligned(16))) C_GARBAGE1[4] = {0x00001337, 0x00001337, 0x00001337, 0x00001337};
+static const u32 __attribute__((aligned(16))) C_GARBAGE2[4] = {0xDEADBEEF, 0xDEADBEEF, 0xDEADBEEF, 0xDEADBEEF};
+static const u32 __attribute__((aligned(16))) C_INCREASING[4] = {0x3F800000, 0x40000000, 0x40400000, 0x40800000};
+static const u32 __attribute__((aligned(16))) C_DECREASING[4] = {0x40800000, 0x40400000, 0x40000000, 0x3F800000};
+
+static u32 *const CVF_ZERO         = (u32 *)(vu1_mem + 0x0000);
+static u32 *const CVF_NEGZERO      = (u32 *)(vu1_mem + 0x0010);
+static u32 *const CVF_MAX          = (u32 *)(vu1_mem + 0x0020);
+static u32 *const CVF_MIN          = (u32 *)(vu1_mem + 0x0030);
+static u32 *const CVF_MAX_MANTISSA = (u32 *)(vu1_mem + 0x0040);
+static u32 *const CVF_MAX_EXP      = (u32 *)(vu1_mem + 0x0050);
+static u32 *const CVF_MIN_EXP      = (u32 *)(vu1_mem + 0x0060);
+static u32 *const CVF_ONE          = (u32 *)(vu1_mem + 0x0070);
+static u32 *const CVF_NEGONE       = (u32 *)(vu1_mem + 0x0080);
+static u32 *const CVF_GARBAGE1     = (u32 *)(vu1_mem + 0x0090);
+static u32 *const CVF_GARBAGE2     = (u32 *)(vu1_mem + 0x00A0);
+static u32 *const CVF_INCREASING   = (u32 *)(vu1_mem + 0x00B0);
+static u32 *const CVF_DECREASING   = (u32 *)(vu1_mem + 0x00C0);
+
+static void setup_vf_constants() {
+	*(vu128 *)CVF_ZERO = *(vu128 *)C_ZERO;
+	*(vu128 *)CVF_NEGZERO = *(vu128 *)C_NEGZERO;
+	*(vu128 *)CVF_MAX = *(vu128 *)C_MAX;
+	*(vu128 *)CVF_MIN = *(vu128 *)C_MIN;
+	*(vu128 *)CVF_MAX_MANTISSA = *(vu128 *)C_MAX_MANTISSA;
+	*(vu128 *)CVF_MAX_EXP = *(vu128 *)C_MAX_EXP;
+	*(vu128 *)CVF_MIN_EXP = *(vu128 *)C_MIN_EXP;
+	*(vu128 *)CVF_ONE = *(vu128 *)C_ONE;
+	*(vu128 *)CVF_NEGONE = *(vu128 *)C_NEGONE;
+	*(vu128 *)CVF_GARBAGE1 = *(vu128 *)C_GARBAGE1;
+	*(vu128 *)CVF_GARBAGE2 = *(vu128 *)C_GARBAGE2;
+	*(vu128 *)CVF_INCREASING = *(vu128 *)C_INCREASING;
+	*(vu128 *)CVF_DECREASING = *(vu128 *)C_DECREASING;
+}
+
+class EfuTestRunner : public TestRunner {
+public:
+	EfuTestRunner(int vu) : TestRunner(vu) {
+	}
+
+	typedef VU::LowerOp (*EfuFunctionScalar)(VU::Field field, VU::Reg reg);
+	typedef VU::LowerOp (*EfuFunctionVector)(VU::Reg reg);
+
+	void PerformS_M(const char *name, EfuFunctionScalar func, const char *tname, const u32 *t) {
+		using namespace VU;
+		
+		Reset();
+		
+		WrLoadFloatRegister(DEST_XYZW, VF01, t);
+		
+		//Op
+		Wr(func(FIELD_Z, VF01));
+		Wr(WAITP());
+		Wr(MFP(DEST_XYZW, VF02));
+		
+		Execute();
+
+		printf("  %s %s: ", name, tname); 
+		PrintRegisterField(VF02, FIELD_X, true);
+	}
+	
+	void PerformV_M(const char *name, EfuFunctionVector func, const char *tname, const u32 *t) {
+		using namespace VU;
+		
+		Reset();
+		
+		WrLoadFloatRegister(DEST_XYZW, VF01, t);
+		
+		//Op
+		Wr(func(VF01));
+		Wr(WAITP());
+		Wr(MFP(DEST_XYZW, VF02));
```

Comment:
Might be worth while to (maybe for just MFP) to check the DEST_ broadcast behavior works, for example to ensure DEST_Y doesn't zero x or something.  Maybe that could be saved for a separated dedicated test, but sometimes things like those cause problems.

-[Unknown]